benchmark/bench_CQRRP/QR_speed_comp.cc

```diff
@@ -0,0 +1,178 @@
+    R(n * n, 0.0),
+    Q(m * n, 0.0),
+    QR(m * n, 0.0),
+    tau(n, 0.0),
+    T_mat(n * n, 0.0),
+    D(n, 0.0)
+    {
+        row = m;
+        col = n;
+    }
+};
+
+// Re-generate and clear data
+template <typename T, typename RNG>
+static void data_regen(RandLAPACK::gen::mat_gen_info<T> m_info, 
+                                        QR_benchmark_data<T> &all_data, 
+                                        RandBLAS::RNGState<RNG> &state,
+                                        int zero_Q) {
+
+    auto state_const = state;
+    RandLAPACK::gen::mat_gen(m_info, all_data.A.data(), state_const);
+    std::fill(all_data.R.begin(), all_data.R.end(), 0.0);
+    std::fill(all_data.tau.begin(), all_data.tau.end(), 0.0);
+    state_const = state;
+    if (zero_Q) {
+        RandLAPACK::gen::mat_gen(m_info, all_data.A_cpy.data(), state_const);
+    }
+}
+
+template <typename T, typename RNG>
+static void call_all_algs(
+    RandLAPACK::gen::mat_gen_info<T> m_info,
+    int64_t numruns,
+    int64_t n,
+    QR_benchmark_data<T> &all_data,
+    RandBLAS::RNGState<RNG> &state,
+    std::string output_filename) {
+
+    auto m        = all_data.row;
+    auto tol      = all_data.tolerance;
+
+    int64_t tsize = 0;
+
+    // timing vars
+    long dur_geqrf       = 0;
+    long dur_geqr        = 0;
+    long dur_geqr_ungqr  = 0;
+    long dur_cholqr      = 0;
+    long dur_cholqr_orhr = 0;
+    
+    // Making sure the states are unchanged
+    auto state_gen = state;
+
+    for (int i = 0; i < numruns; ++i) {
+        printf("Iteration %d start.\n", i);
+        // Testing GEQRF
+        auto start_geqrf = high_resolution_clock::now();
+        lapack::geqrf(m, n, all_data.A.data(), m, all_data.tau.data());
+        auto stop_geqrf = high_resolution_clock::now();
+        dur_geqrf = duration_cast<microseconds>(stop_geqrf - start_geqrf).count();
+
+        state_gen = state;
+        data_regen(m_info, all_data, state_gen, 0);
+
+        // Testing GEQR
+        auto start_geqr = high_resolution_clock::now();
+#if !defined(__APPLE__)
+        lapack::geqr(m, n, all_data.A.data(), m,  all_data.tau.data(), -1);
+        tsize = (int64_t) all_data.tau[0]; 
+        all_data.tau.resize(tsize);
+        lapack::geqr(m, n, all_data.A.data(), m, all_data.tau.data(), tsize);
+#endif
+        auto stop_geqr = high_resolution_clock::now();
+        dur_geqr = duration_cast<microseconds>(stop_geqr - start_geqr).count();
+
+        state_gen = state;
+        data_regen(m_info, all_data, state_gen, 0);
+
+        // Testing GEQR + UNGQR
+        auto start_geqr_ungqr = high_resolution_clock::now();
+#if !defined(__APPLE__)
+        lapack::geqr(m, n, all_data.A.data(), m,  all_data.tau.data(), -1);
+        tsize = (int64_t) all_data.tau[0]; 
+        all_data.tau.resize(tsize);
+        lapack::geqr(m, n, all_data.A.data(), m, all_data.tau.data(), tsize);
+        lapack::ungqr(m, n, n, all_data.A.data(), m, all_data.tau.data());
+#endif
```

Comment:
Please make this one #if directive. Define variables in an #else block if needed.

Reply:
Done.